File: examples/csharp/IndexStore.cs
// 胶水 C# 示例：同一个工具、同一套头尾标记，在 C# 里也能用。
// 证明「引擎逻辑与被管语言无关」——C# 用的就是与 Java 相同的 // 行注释标记。
// 候选一：SortedDictionary（红黑树）  候选二：有序 List + 二分（B+ 树简化）

using System;
using System.Collections.Generic;

class IndexStore
{
    // 候选一：红黑树
    private readonly SortedDictionary<int, string> rb = new();

    // 候选二：B+ 树（简化：有序 List + 二分）
    private readonly List<int> bpKeys = new();
    private readonly List<string> bpVals = new();

    public void Put(int key, string val)
    {
        rb[key] = val;
        int i = bpKeys.BinarySearch(key);
        if (i >= 0) bpVals[i] = val;
        else { bpKeys.Insert(~i, key); bpVals.Insert(~i, val); }
    }

    public string? Get(int key)
    {
        // glue:begin cs_lookup
        return rb.TryGetValue(key, out var v) ? v : null;
        // glue:end
    }

    public static void Main()
    {
        var s = new IndexStore();
        int n = 200_000;
        for (int i = 0; i < n; i++) s.Put(i, "v" + i);

        long best = long.MaxValue;
        for (int round = 0; round < 5; round++)
        {
            long t0 = System.Diagnostics.Stopwatch.GetTimestamp();
            long acc = 0;
            for (int i = 0; i < n; i++) if (s.Get(i) != null) acc++;
            long t1 = System.Diagnostics.Stopwatch.GetTimestamp();
            long ms = (t1 - t0) * 1000 / System.Diagnostics.Stopwatch.Frequency;
            if (ms < best) best = ms;
            if (acc != n) throw new Exception("结果不一致: " + acc);
        }
        Console.WriteLine($"20 万次查找，最快一轮: {best} ms");
    }
}

File: examples/demo/IndexStore.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

/**
 * 胶水示例：索引查找到底该用红黑树还是 B+ 树？
 *
 * 这块拿不准，于是用头尾两行注释框起来 —— 中间这一块就是"胶水"。
 * 两个候选实现都登记在 .glue/lookup/ 下，用 glue use 切换，
 * 切换直接改源码文本，运行时没有任何间接层。
 */
public class IndexStore {

    // 候选一：红黑树（JDK TreeMap）
    private final TreeMap<Integer, String> rbtree = new TreeMap<>();

    // 候选二：B+ 树（此处简化为有序数组 + 二分，只为演示切换流程）
    private final ArrayList<Integer> bpKeys = new ArrayList<>();
    private final ArrayList<String> bpVals = new ArrayList<>();

    public void put(int key, String val) {
        rbtree.put(key, val);
        int i = Collections.binarySearch(bpKeys, key);
        if (i >= 0) bpVals.set(i, val);
        else { bpKeys.add(-i - 1, key); bpVals.add(-i - 1, val); }
    }

    public String get(int key) {
        // glue:begin lookup
        int found = Collections.binarySearch(bpKeys, key);
        return (found < 0) ? null : bpVals.get(found);
        // glue:end
    }

    public int size() {
        return rbtree.size();
    }

    public static void main(String[] args) {
        IndexStore s = new IndexStore();
        int n = 200_000;
        for (int i = 0; i < n; i++) s.put(i, "v" + i);

        long best = Long.MAX_VALUE;
        for (int round = 0; round < 5; round++) {
            long t0 = System.nanoTime();
            long acc = 0;
            for (int i = 0; i < n; i++) if (s.get(i) != null) acc++;
            long t1 = System.nanoTime();
            best = Math.min(best, t1 - t0);
            if (acc != n) throw new IllegalStateException("结果不一致: " + acc);
        }
        System.out.printf("20 万次查找，最快一轮: %.2f ms%n", best / 1e6);
    }
}
